File: go/ie/mg.go
package ie

import (
	"regexp"
	"strconv"

	"github.com/matheuslm7/brazil-validator/go/internal/shared"
)

// Verified against the official SEFAZ-MG "Roteiro de Crítica da Inscrição
// Estadual" (mirrored at sintegra.gov.br/Cad_Estados/cad_MG.html), including
// its fully worked example (062.307.904/0081).
//
// Format: A1A2A3 B1B2B3B4B5B6 C1C2 D1D2 (13 digits), where A = município
// code, B = registration number, C = establishment order, D = check digits.
var (
	mgAllowedChars       = regexp.MustCompile(`^[\d./\s]+$`)
	mgFirstDigitWeights  = []int{1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2}
	mgSecondDigitWeights = []int{3, 2, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2}
)

func mgNormalize(value string) string {
	return shared.RemoveNonDigits(value)
}

func mgSumOfDigits(value int) int {
	sum := 0

	for _, c := range strconv.Itoa(value) {
		sum += int(c - '0')
	}

	return sum
}

// mgCalculateFirstDigit equalizes the field widths by inserting a "0" right
// after the município code, then sums the *digits* of each weighted
// product (not the products themselves) before completing to the next
// multiple of ten.
func mgCalculateFirstDigit(base string) int {
	withInsertedZero := base[:3] + "0" + base[3:]

	digitSum := 0

	for i, w := range mgFirstDigitWeights {
		digitSum += mgSumOfDigits(int(withInsertedZero[i]-'0') * w)
	}

	remainder := digitSum % 10

	if remainder == 0 {
		return 0
	}

	return 10 - remainder
}

// mgCalculateSecondDigit uses the original (un-padded) base plus D1, and
// sums the weighted products directly, following the general módulo 11
// remainder rule.
func mgCalculateSecondDigit(base string, firstDigit int) int {
	withFirstDigit := base + strconv.Itoa(firstDigit)

	sum := shared.WeightedSum(withFirstDigit, mgSecondDigitWeights)
	remainder := sum % 11

	if remainder <= 1 {
		return 0
	}

	return 11 - remainder
}

func mgIsValid(value string) bool {
	if !mgAllowedChars.MatchString(value) {
		return false
	}

	v := mgNormalize(value)

	if len(v) != 13 {
		return false
	}

	base := v[:11]
	firstDigit := mgCalculateFirstDigit(base)

	if firstDigit != int(v[11]-'0') {
		return false
	}

	secondDigit := mgCalculateSecondDigit(base, firstDigit)

	return secondDigit == int(v[12]-'0')
}

func mgFormat(value string) string {
	v := mgNormalize(value)

	if len(v) != 13 {
		return value
	}

	return v[:3] + "." + v[3:6] + "." + v[6:9] + "/" + v[9:13]
}

File: php/src/Ie/Mg.php
<?php

declare(strict_types=1);

namespace BrValidator\Ie;

use BrValidator\Shared\Digits;
use BrValidator\Shared\Mod11;

/**
 * Verified against the official SEFAZ-MG "Roteiro de Critica da Inscricao
 * Estadual" (mirrored at sintegra.gov.br/Cad_Estados/cad_MG.html),
 * including its fully worked example (062.307.904/0081).
 *
 * Format: A1A2A3 B1B2B3B4B5B6 C1C2 D1D2 (13 digits), where A = municipio
 * code, B = registration number, C = establishment order, D = check
 * digits.
 *
 * Not part of the public API - use BrValidator\Ie.
 */
final class Mg
{
    private const ALLOWED_CHARS = '/^[\d.\/\s]+$/';
    private const FIRST_DIGIT_WEIGHTS = [1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2];
    private const SECOND_DIGIT_WEIGHTS = [3, 2, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2];

    private function __construct()
    {
    }

    public static function normalize(string $value): string
    {
        return Digits::removeNonDigits($value);
    }

    private static function sumOfDigits(int $value): int
    {
        $sum = 0;

        foreach (str_split((string) $value) as $char) {
            $sum += (int) $char;
        }

        return $sum;
    }

    /**
     * Equalizes the field widths by inserting a "0" right after the
     * municipio code, then sums the *digits* of each weighted product
     * (not the products themselves) before completing to the next
     * multiple of ten.
     */
    private static function calculateFirstDigit(string $base): int
    {
        $withInsertedZero = substr($base, 0, 3) . '0' . substr($base, 3);

        $digitSum = 0;

        foreach (self::FIRST_DIGIT_WEIGHTS as $i => $weight) {
            $digitSum += self::sumOfDigits((int) $withInsertedZero[$i] * $weight);
        }

        $remainder = $digitSum % 10;

        if ($remainder === 0) {
            return 0;
        }

        return 10 - $remainder;
    }

    /**
     * Uses the original (un-padded) base plus D1, and sums the weighted
     * products directly, following the general modulo 11 remainder rule.
     */
    private static function calculateSecondDigit(string $base, int $firstDigit): int
    {
        $withFirstDigit = $base . (string) $firstDigit;

        $sum = Mod11::weightedSum($withFirstDigit, self::SECOND_DIGIT_WEIGHTS);
        $remainder = $sum % 11;

        if ($remainder <= 1) {
            return 0;
        }

        return 11 - $remainder;
    }

    public static function isValid(string $value): bool
    {
        if (!preg_match(self::ALLOWED_CHARS, $value)) {
            return false;
        }

        $v = self::normalize($value);

        if (strlen($v) !== 13) {
            return false;
        }

        $base = substr($v, 0, 11);
        $firstDigit = self::calculateFirstDigit($base);

        if ($firstDigit !== (int) $v[11]) {
            return false;
        }

        $secondDigit = self::calculateSecondDigit($base, $firstDigit);

        return $secondDigit === (int) $v[12];
    }

    public static function format(string $value): string
    {
        $v = self::normalize($value);

        if (strlen($v) !== 13) {
            return $value;
        }

        return substr($v, 0, 3) . '.' . substr($v, 3, 3) . '.' . substr($v, 6, 3) . '/' . substr($v, 9, 4);
    }
}
